File: app/src/main/java/com/lugo/manueln/socialapp/Presenters/PresenterPosts.java
package com.lugo.manueln.socialapp.Presenters;

import android.support.v4.app.FragmentActivity;

import com.lugo.manueln.socialapp.Interfaces.InterPost;
import com.lugo.manueln.socialapp.Interactors.ModulePosts;
import com.lugo.manueln.socialapp.objects.Post;

import java.util.List;

public class PresenterPosts implements InterPost.presenter {

    private InterPost.view view;
    private InterPost.interactor model;

    public PresenterPosts(InterPost.view view){
        this.view=view;
        model=new ModulePosts(this);
    }
    @Override
    public void loadRecyclerPostPresenter(FragmentActivity activity) {

        if(model!=null){

            model.getPosts(activity);
        }
    }


    @Override
    public void sendPostPresenter(List<Post> miListaPost) {

        if(view!=null){

            view.showRecycler(miListaPost);
        }

    }

    @Override
    public void sendErrorPostPresenter(String error) {
        if(view!=null){

            view.errorLoadRecyclerPost(error);
        }

    }

    @Override
    public void newPostCompleteFragmentPresenter(int idPost,FragmentActivity fragmentActivity) {

        if(model!=null){

            model.newPostCompleteFragment(idPost,fragmentActivity);
        }

    }

    @Override
    public void newProfileFragmentPresenter(String userName, FragmentActivity fragmentActivity) {

        if(model!=null){

            model.newProfileFragment(userName,fragmentActivity);
        }
    }
}

File: app/src/main/java/com/lugo/manueln/socialapp/presentation/Views/Posts/Presenter/PresenterPosts.kt
package com.lugo.manueln.socialapp.presentation.Views.Posts.Presenter

import com.lugo.manueln.socialapp.domain.Post
import com.lugo.manueln.socialapp.presentation.Views.Posts.PostContract
import com.lugo.manueln.socialapp.usecases.Posts.GetListPosts
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers

class PresenterPosts(private var view: PostContract.view?,private val getListPosts: GetListPosts) : PostContract.presenter {


    override fun loadRecyclerPostPresenter() {

        getListPosts.invoke().subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Observer<List<Post>> {
                    override fun onComplete() {

                    }

                    override fun onSubscribe(d: Disposable?) {

                    }

                    override fun onNext(listPost: List<Post>) {

                        view?.showRecycler(listPost)
                    }

                    override fun onError(e: Throwable) {


                        view?.errorLoadRecyclerPost(e.message)
                    }
                })


    }


    override fun sendPostPresenter(miListaPost: List<Post>) {

        if (isViewAttach()) {

            view!!.showRecycler(miListaPost)
        }

    }

    override fun sendErrorPostPresenter(error: String?) {
        if (isViewAttach()) {

            view!!.errorLoadRecyclerPost(error)
        }

    }

    override fun detachView() {
        view= null
    }

    override fun isViewAttach(): Boolean=view!=null
}
